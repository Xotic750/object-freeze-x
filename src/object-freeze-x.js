import isPimitive from 'is-primitive-x';
import isTypedArray from 'is-typed-array';

const nativeFreeze = {}.constructor.freeze;

const assertTypedArray = function assertTypedArray(obj) {
  if (isTypedArray(obj) && obj.byteLength !== 0) {
    throw new TypeError('Cannot freeze array buffer views with elements');
  }

  return obj;
};

export const patchedFreeze = function freeze(obj) {
  return isPimitive(obj) ? obj : nativeFreeze(assertTypedArray(obj));
};

// fake
export const implementation = function freeze(obj) {
  return assertTypedArray(obj);
};

/**
 * This method method freezes an object. A frozen object can no longer be changed; freezing an
 * object prevents new properties from being added to it, existing properties from being removed,
 * prevents changing the enumerability, configurability, or writability of existing properties,
 * and prevents the values of existing properties from being changed. In addition, freezing an
 * object also prevents its prototype from being changed. It returns the same object that
 * was passed in.
 *
 * @param {*} obj - The object to freeze.
 * @returns {*} The object that was passed to the function..
 */
const freeze = typeof nativeFreeze === 'function' ? patchedFreeze : implementation;

export default freeze;
