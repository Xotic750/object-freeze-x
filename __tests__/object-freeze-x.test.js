import $P, {implementation as $I} from '../src/object-freeze-x';

[$I, $P].forEach((freeze, testNum) => {
  describe(`freeze ${testNum}`, function() {
    it('is a function', function() {
      expect.assertions(1);
      expect(typeof freeze).toBe('function');
    });

    it('should throw when typed array has elements', function() {
      expect.assertions(2);
      const uint8 = new Uint8Array(0);
      expect(freeze(uint8)).toBe(uint8);
      const uint81 = new Uint8Array(1);
      expect(function() {
        freeze(uint81);
      }).toThrowErrorMatchingSnapshot();
    });

    it('should return primitives', function() {
      expect.assertions(6);
      expect(freeze()).toBeUndefined();
      expect(freeze(undefined)).toBeUndefined();
      expect(freeze(null)).toBeNull();
      expect(freeze(1)).toBe(1);
      expect(freeze(true)).toBe(true);
      expect(freeze('x')).toBe('x');
    });

    it('should return objects', function() {
      expect.assertions(4);
      const array = [];
      expect(freeze(array)).toBe(array);
      const object = {};
      expect(freeze(object)).toBe(object);
      const date = new Date();
      expect(freeze(date)).toBe(date);
      const rx = /none/;
      expect(freeze(rx)).toBe(rx);
    });
  });
});
