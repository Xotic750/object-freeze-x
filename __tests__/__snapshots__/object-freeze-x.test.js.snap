// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`freeze 0 should throw when typed array has elements 1`] = `"Cannot freeze array buffer views with elements"`;

exports[`freeze 1 should throw when typed array has elements 1`] = `"Cannot freeze array buffer views with elements"`;
